tests/unit/meta/test_command.py

```diff
@@ -108,6 +143,32 @@ def test_command(self):
         )
 
         self.assertThat(cmd.command, Equals("foo"))
+        self.assertThat(self.fake_logger.output, Equals(""))
+
+    def test_command_with_dollar_snap_and_does_not_match_snapd_pattern(self):
+        _create_file(os.path.join(self.path, "foo"))
+        cmd = command.Command(
+            app_name="foo", command_name="command", command="$SNAP/foo !option"
+        )
+        cmd.prime_command(
+            can_use_wrapper=True, massage_command=True, prime_dir=self.path
+        )
+
+        self.assertThat(cmd.command, Equals("command-foo.wrapper"))
+        self.assertThat(
+            self.fake_logger.output.strip(),
+            Equals(
+                "Command '$SNAP/foo !option' has been changed to 'foo !option' "
```

Comment:
This doesn't look quite right due to the mixed messaging `($ is allowed, ! is not, but  you lost $ and kept !)`.  But looking at it in this context is helpful :D

When use of "$SNAP" is detected, I would maybe just independently indicate that it's being dropped because it's not required:
```
"Stripping '$SNAP/' from command '<command>'.  Relative paths are assumed to be relative to $SNAP."
```

A suggested format for wrappers due to special characters:
```
Snapcraft created a wrapper script to execute '<command>' .  Commands may only consist of alphanumeric characters, spaces, and the following special characters: `/ . _ # : $ -'` and the configured command contains unsupported characters.  To resolve this issue for you automatically, snapcraft configured a wrapper script '<relative path>' to execute '<wrapped command>'.  To remove the need for this wrapper script, you must remove unsupported characters from the command.
```